fix graph deepcopy to propagate output type

migeed-z opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61790
* #61776
* **#61747**
* #61310

Differential Revision: [D29737565](https://our.internmc.facebook.com/intern/diff/D29737565)

migeed-z committed:
fix graph deepcopy to propagate output type

[ghstack-poisoned]

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -358,6 +358,9 @@ def __deepcopy__(self, memo=None) -> 'Graph':
         g.output(output_val)
+        for n, m in zip(g.nodes, self.nodes):
+            if n.op == 'output' and hasattr(m, 'type'):
+                n.type = m.type
```
How about instead we do this:
1) Make `Graph.graph_copy` have a flag `return_output_node=False` that, when `True`, makes it so that the method returns a tuple of (new output value, old output node)
2) Call `graph_copy` with `return_output_node=True` at line 359 here
3) Pass the type from the original output node to the `output` call at line 360

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29737565).

migeed-z committed:
Update on "fix graph deepcopy to propagate output type"


Differential Revision: [D29737565](https://our.internmc.facebook.com/intern/diff/D29737565)

[ghstack-poisoned]

migeed-z committed:
Update on "fix graph deepcopy to propagate output type"


Differential Revision: [D29737565](https://our.internmc.facebook.com/intern/diff/D29737565)

[ghstack-poisoned]

migeed-z committed:
Update on "fix graph deepcopy to propagate output type"


Differential Revision: [D29737565](https://our.internmc.facebook.com/intern/diff/D29737565)

[ghstack-poisoned]

migeed-z committed:
Update on "fix graph deepcopy to propagate output type"


Differential Revision: [D29737565](https://our.internmc.facebook.com/intern/diff/D29737565)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29737565).

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29737565).

migeed-z committed:
Update on "fix graph deepcopy to propagate output type"


Differential Revision: [D29737565](https://our.internmc.facebook.com/intern/diff/D29737565)

[ghstack-poisoned]

jamesr66a commented on torch/fx/graph.py:
```diff
@@ -356,8 +356,13 @@ def __deepcopy__(self, memo=None) -> 'Graph':
+        if hasattr(old_output_val, 'type'):
+            g.output(output_val, old_output_val.type)
+        else:
+            g.output(output_val)
```
You can replace this whole `if/else` with `g.output(output_val, type_expr=getattr(old_output_val, 'type', None))

jamesr66a reviewed (approved):
Great!

migeed-z committed:
Update on "fix graph deepcopy to propagate output type"


Differential Revision: [D29737565](https://our.internmc.facebook.com/intern/diff/D29737565)

[ghstack-poisoned]

migeed-z committed:
Update on "fix graph deepcopy to propagate output type"


Differential Revision: [D29737565](https://our.internmc.facebook.com/intern/diff/D29737565)

[ghstack-poisoned]

migeed-z committed:
Update on "fix graph deepcopy to propagate output type"


Differential Revision: [D29737565](https://our.internmc.facebook.com/intern/diff/D29737565)

[ghstack-poisoned]

migeed-z committed:
Update on "fix graph deepcopy to propagate output type"


Differential Revision: [D29737565](https://our.internmc.facebook.com/intern/diff/D29737565)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29737565).
